src/test/java/coresearch/cvurl/io/request/CVurlProxySelectorTest.java

```diff
@@ -0,0 +1,155 @@
+        // given
+        final List<Proxy> expectedProxies = List.of(makeExpectedHttpProxy(PROXY_HOST, PROXY_PORT),
+                makeExpectedHttpProxy(PROXY_HOST + 2, PROXY_PORT));
+        selector.addProxy(REQUEST_URL, CVurlProxy.of(Proxy.Type.HTTP, PROXY_HOST, PROXY_PORT));
+        selector.addProxy(REQUEST_URL, CVurlProxy.of(Proxy.Type.HTTP, PROXY_HOST + 2, PROXY_PORT));
+
+        // when
+        final List<Proxy> proxies = selector.select(REQUEST_URI);
+
+        //then
+        Assertions.assertEquals(expectedProxies, proxies);
+    }
+
+    @Test
+    public void whenProxySelectorSpecifiedAndNoPerRequestProxy_returnProxiesFromProxySelector() {
+        // given
+        final List<Proxy> expectedProxies = List.of(makeExpectedHttpProxy(PROXY_HOST, PROXY_PORT));
+        selector = new CVurlProxySelector(new ProxySelectorMock(expectedProxies));
+
+        // when
+        final List<Proxy> proxies = selector.select(REQUEST_URI);
+
+        // then
+        Assertions.assertEquals(expectedProxies, proxies);
+    }
+
+    @Test
+    public void whenProxySelectorSpecifiedWithPerRequestProxy_returnPerRequestProxy() {
+        // given
+        selector = new CVurlProxySelector(new ProxySelectorMock(List.of(makeExpectedHttpProxy(PROXY_HOST, PROXY_PORT))));
+        selector.addProxy(REQUEST_URL, CVurlProxy.of(Proxy.Type.HTTP, PROXY_HOST + 2, PROXY_PORT));
+
+        // when
+        final List<Proxy> proxies = selector.select(REQUEST_URI);
+
+        // then
+        Assertions.assertEquals(1, proxies.size());
+        Assertions.assertEquals(makeExpectedHttpProxy(PROXY_HOST + 2, PROXY_PORT), proxies.get(0));
+    }
+
+    @Test
+    public void whenNoProxySelectorAndNoPerRequestProxySpecified_returnDefaultProxy() {
+        // when
+        final List<Proxy> proxies = selector.select(REQUEST_URI);
+
+        // then
+        Assertions.assertEquals(ProxySelector.getDefault().select(REQUEST_URI), proxies);
+    }
+
+    @Test
+    public void whenSelectForUriCalledTwice_returnDefaultProxyForThisUriForSecondCallForSameUri() {
+        // given
+        selector.addProxy(REQUEST_URL, CVurlProxy.of(Proxy.Type.HTTP, PROXY_HOST, PROXY_PORT));
+        selector.select(REQUEST_URI);
+
+        // when
+        final List<Proxy> proxiesSecond = selector.select(REQUEST_URI);
+
+        // then
+        Assertions.assertEquals(ProxySelector.getDefault().select(REQUEST_URI), proxiesSecond);
+    }
+
+    @Test
+    public void whenNoProxyIsUsed_returnNoProxy() {
+        // given
+        selector.addProxy(REQUEST_URL, CVurlProxy.noProxy());
+
+        // when
+        final List<Proxy> proxies = selector.select(REQUEST_URI);
+
+        // then
+        Assertions.assertEquals(List.of(Proxy.NO_PROXY), proxies);
+    }
+
+    private Proxy makeExpectedHttpProxy(String host, int port) {
+        return new Proxy(Proxy.Type.HTTP, new InetSocketAddress(host, port));
+    }
+
+    private static final class ProxySelectorMock extends ProxySelector {
```

Comment:
It could be used not only for this test. I recommend moving it to a separate class in test helpers package

Reply:
Done.